src/accel/LocalTransporter.cc

```diff
@@ -124,6 +128,38 @@ void LocalTransporter::SetEventId(int id)
     track_counter_ = 0;
 }
 
+//---------------------------------------------------------------------------//
+/*!
+ * Set the event ID and reseed the Celeritas RNG at the start of an event.
+ */
+void LocalTransporter::InitializeEvent(int id)
+{
+    CELER_EXPECT(*this);
+    CELER_EXPECT(id >= 0);
+
+    this->SetEventId(id);
+
+    if (G4Threading::IsMultithreadedApplication()
+        && !G4MTRunManager::SeedOncePerCommunication())
+    {
+        // Since Geant4 schedules events dynamically, reseed the Celeritas RNGs
+        // using the Geant4 event ID for reproducibility.
+        if (celeritas::device())
+        {
+            auto const& state
+                = dynamic_cast<CoreState<MemSpace::device> const&>(
+                    step_->state());
+            detail::reseed_rng(rng_->device_ref(), state.ref().rng, id);
+        }
+        else
+        {
+            auto const& state = dynamic_cast<CoreState<MemSpace::host> const&>(
+                step_->state());
+            detail::reseed_rng(rng_->host_ref(), state.ref().rng, id);
+        }
+    }
```

Comment:
Let's perhaps add a `reseed` method to the stepper? It'll make this a single line, remove the const casting, and hide the tempating.